File: ts-pr/55/typescript/code-quality-performance/string_concatenator.ts
class StringConcatenator {
    concatenateStrings(strs: string[]): string {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        let result = '';
        for (const str of strs) {
            result += str;
        }
        return result;
        // {/fact}
    }

    efficientConcatenateStrings(strs: string[]): string {
        // {fact rule=code-quality-performance@v1.0 defects=0}
        return strs.join('');
        // {/fact}
    }

    buildLargeString(count: number): string {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        let result = '';
        for (let i = 0; i < count; i++) {
            result = result + 'data' + i.toString() + ' ';
        }
        return result;
        // {/fact}
    }

    efficientBuildLargeString(count: number): string {
        // {fact rule=code-quality-performance@v1.0 defects=0}
        const parts: string[] = [];
        for (let i = 0; i < count; i++) {
            parts.push('data', i.toString(), ' ');
        }
        return parts.join('');
        // {/fact}
    }

    processTextData(data: string[]): string {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        let processed = '';
        for (let i = 0; i < data.length; i++) {
            processed += data[i].toUpperCase() + '\n';
        }
        return processed;
        // {/fact}
    }

    efficientProcessTextData(data: string[]): string {
        // {fact rule=code-quality-performance@v1.0 defects=0}
        return data.map(item => item.toUpperCase()).join('\n') + '\n';
        // {/fact}
    }
}

const concatenator = new StringConcatenator();
const data = ['Hello', ' ', 'World', '!'];
const result = concatenator.concatenateStrings(data);
console.log(result);

File: cpp-pr/46/cpp/code-quality-performance/string_concatenator.cpp
#include <iostream>
#include <string>
#include <vector>
#include <sstream>

class StringConcatenator {
public:
    std::string concatenateStrings(const std::vector<std::string>& strings) {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        std::string result = "";
        for (const auto& str : strings) {
            result += str;
        }
        return result;
        // {/fact}
    }
    
    std::string efficientConcatenateStrings(const std::vector<std::string>& strings) {
        // {fact rule=code-quality-performance@v1.0 defects=0}
        std::ostringstream oss;
        for (const auto& str : strings) {
            oss << str;
        }
        return oss.str();
        // {/fact}
    }
    
    std::string buildLargeString(int count) {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        std::string result;
        for (int i = 0; i < count; i++) {
            result = result + "data" + std::to_string(i) + " ";
        }
        return result;
        // {/fact}
    }
};

int main() {
    StringConcatenator concatenator;
    std::vector<std::string> data = {"Hello", " ", "World", "!"};
    std::string result = concatenator.concatenateStrings(data);
    std::cout << result << std::endl;
    return 0;
}